src/ipc-wrapper/client.js

```diff
@@ -0,0 +1,113 @@
+// @ts-check
+import { createClient } from 'rpc-reflector'
+import pTimeout from 'p-timeout'
+import { MANAGER_CHANNEL_ID, MAPEO_IPC_ID, SubChannel } from './sub-channel.js'
+
+const CLOSE = Symbol('close')
+
+/**
+ * @param {import('./sub-channel.js').MessagePortLike} messagePort
+ * @returns {import('rpc-reflector/client.js').ClientApi<import('../mapeo-manager.js').MapeoManager>}
+ */
+export function createMapeoClient(messagePort) {
+  /** @type {Map<import('../types.js').ProjectPublicId, { instance: import('rpc-reflector/client.js').ClientApi<import('../mapeo-project.js').MapeoProject>, channel: SubChannel }>} */
+  const existingProjectClients = new Map()
+
+  const mapeoIpcChannel = new SubChannel(messagePort, MAPEO_IPC_ID)
+  mapeoIpcChannel.start()
+
+  const managerChannel = new SubChannel(messagePort, MANAGER_CHANNEL_ID)
+  managerChannel.start()
+
+  /** @type {import('rpc-reflector').ClientApi<import('../mapeo-manager.js').MapeoManager>} */
+  const managerClient = createClient(managerChannel)
+
+  const client = new Proxy(managerClient, {
+    get(target, prop, receiver) {
+      if (prop === CLOSE) {
+        return () => {
+          mapeoIpcChannel.close()
+          managerChannel.close()
+          createClient.close(managerClient)
+        }
+      }
+
+      if (prop === 'getProject') {
+        return createProjectClient
+      }
+
+      return Reflect.get(target, prop, receiver)
+    },
+  })
+
+  return client
+
+  /**
+   * @param {import('../types.js').ProjectPublicId} projectPublicId
+   * @returns {Promise<import('rpc-reflector/client.js').ClientApi<import('../mapeo-project.js').MapeoProject>>}
+   */
+  async function createProjectClient(projectPublicId) {
+    const existingClient = existingProjectClients.get(projectPublicId)
+
+    if (existingClient) return existingClient.instance
+
+    // Wait for the server to confirm that it could retrieve the desired project
+    await pTimeout(
```

Comment:
Unfortunately another race condition here. I believe this test will fail:

```js
const projectId = await client.createProject()
const [projectInstance1, projectInstance2] = await Promise.all([
  client.getProject(projectId),
  client.getProject(projectId)
])
t.is(projectInstance1, projectInstance2)
```

One way to solve this it with a deferred promise (I use `p-defer`). 

```js
  if (existingClient) return existingClient.deferredInstance
  const deferred = pDefer()
  existingProjectClients.set(projectPublicId, deferred.promise)

  //...

  deferred.resolve(projectClient)
```

Will also need to reject the deferred promise in the catch, check with this test:

```js
const results = await Promise.allSettled([
  client.getProject('aaaa'),
  client.getProject('aaaa')
])
t.alike(results.map(({ status }) => status), ['rejected', 'rejected'])
```